### PR TITLE
🍰 feature/architecture

### PR DESCRIPTION
**Pull request by Olderestin**
## What changed?     
I have changed architectury of project
Update Docker container build
Make docstrings
Add Poetry to the project

## Checklist before requesting a review     
- [x] I have performed a self-review of my code
- [x] I have run pre-commit over my code
- [x] Code is functional on my machine(Windows 10/Debian GNU/Linux 11 (bullseye))

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,5 +3,5 @@ venv
 __pycache__
 *.pyc
 .vscode
-coin_flip_bot.db
+user.db
```
Comment (memhard) [inline .gitignore]: consider the possibility ignore all files in storage root

File: bot/__main__.py [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
 
-from handlers import start, flip, stats, about
-from config import settings
-from models import UserDatabase
+from bot.handlers import start, flip, stats, about
+from bot.config import settings
+from bot.db import UserDatabase
 
-logger.add("app.log", rotation="500 MB", level="TRACE")
+logger.add("storage/app.log", rotation="500 MB", level="TRACE")
```
Comment (memhard) [inline bot/__main__.py]: consider the possibility of using Pathlib variable, not a string just

File: bot/db.py [hunk 3]
```diff
@@ -2,19 +2,34 @@
...
 
 
 class UserDatabase():
+    """
+    A class for interacting with a SQLite database to store user statistics.
+
+    param db_path: The path to sqlite3 database file.
+
+    attribute db_path: The path to connect to sqlite3 database.
+    """
```
Comment (memhard) [inline bot/db.py]: why describe variables in the class description?

File: bot/db.py [hunk 4]
```diff
@@ -23,6 +38,15 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
 
     def create_user_table(self):
+        """
+        Creates the 'user' table in the database if it doesn't exist.
+
+        column user_id: Primary key representing the user.
+        column flips_count: Count of coin flips for the user with a default value of 0.
+        column heads_count: Count of 'heads' outcomes for the user with a default value of 0.
+        column tails_count: Count of 'tails' outcomes for the user with a default value of 0.
+        """
```
Comment (memhard) [inline bot/db.py]: unnecessary information

File: bot/db.py [hunk 5]
```diff
@@ -48,6 +84,12 @@ def make_flip(self, user_id: int):
         return result
 
     def get_stats(self, user_id: int):
+        """
+        Retrieves user statistics from the database and returns it.
+        
+        param user_id: The ID of the user.
+        """
```
Comment (memhard) [inline bot/db.py]: where annotation + outdated docs

File: bot/db.py [hunk 6]
```diff
@@ -2,19 +2,34 @@
...
     def __init__(self, db_path: Path = settings.DATABASE_URL):
         self.db_path = db_path
 
     def __enter__(self):
-        self.conn = sqlite3.connect(self.db_path)
+        """
+        Called when entering a 'with' statement. Connects to the database.
+        """
+
+        self.conn = sqlite3.connect(self.db_path / "user.db")
```
Comment (memhard) [inline bot/db.py]: consider the possibility less use connections to the database

File: bot/db.py [hunk 7]
```diff
@@ -2,19 +2,34 @@
...
         self.cursor = self.conn.cursor()
         return self
     
     def __exit__(self, exc_type, exc_val, exc_tb):
+        """
+        Called when exit a 'with' statement. Commits changes if no exceptions occurred.
+        """
+
         try:
```
Comment (memhard) [inline bot/db.py:38]: consider the possibility the closing cursors, not connections

File: bot/db.py [hunk 8]
```diff
@@ -33,9 +57,21 @@ def create_user_table(self):
...
     def add_user(self, user_id: int):
+        """
+        Adds a user to the 'user' table if they don't already exist.
+
+        param user_id: The ID of the user to be added.
+        """
+
         self.cursor.execute('INSERT OR IGNORE INTO user (user_id) VALUES (?)', (user_id,))
 
     def make_flip(self, user_id: int):
```
Comment (memhard) [inline bot/db.py]: where annotation

File: bot/db.py [hunk 9]
```diff
@@ -33,9 +57,21 @@ def create_user_table(self):
     ''')
         
     def add_user(self, user_id: int):
```
Comment (memhard) [inline bot/db.py]: where annotation

File: bot/db.py [hunk 10]
```diff
@@ -23,6 +38,15 @@ def __exit__(self, exc_type, exc_val, exc_tb):
             del self.cursor
 
     def create_user_table(self):
```
Comment (memhard) [inline bot/db.py]: where annotation

File: bot/db.py [hunk 11]
```diff
@@ -2,19 +2,34 @@
...
-        self.conn = sqlite3.connect(self.db_path)
+        """
+        Called when entering a 'with' statement. Connects to the database.
+        """
+
+        self.conn = sqlite3.connect(self.db_path / "user.db")
         self.cursor = self.conn.cursor()
         return self
     
     def __exit__(self, exc_type, exc_val, exc_tb):
```
Comment (memhard) [inline bot/db.py]: where annotation

File: bot/db.py [hunk 12]
```diff
@@ -2,19 +2,34 @@
...
+    A class for interacting with a SQLite database to store user statistics.
+
+    param db_path: The path to sqlite3 database file.
+
+    attribute db_path: The path to connect to sqlite3 database.
+    """
     def __init__(self, db_path: Path = settings.DATABASE_URL):
         self.db_path = db_path
 
     def __enter__(self):
```
Comment (memhard) [inline bot/db.py]: where annotation

File: bot/db.py [hunk 13]
```diff
@@ -2,19 +2,34 @@
...
 
 class UserDatabase():
+    """
+    A class for interacting with a SQLite database to store user statistics.
+
+    param db_path: The path to sqlite3 database file.
+
+    attribute db_path: The path to connect to sqlite3 database.
+    """
     def __init__(self, db_path: Path = settings.DATABASE_URL):
```
Comment (memhard) [inline bot/db.py]: where annotation

File: dockerfile [hunk 14]
```diff
@@ -3,9 +3,13 @@ FROM python:3.10.9
...
-RUN mkdir -p /src
-WORKDIR /src
+COPY requirements.txt /app/temp/requirements.txt
 
 RUN pip install --upgrade pip
-RUN pip install -r /temp/requirements.txt
\ No newline at end of file
+RUN pip install -r /app/temp/requirements.txt
+
+COPY . /app/bot
```
Comment (memhard) [inline dockerfile]: COPY . /app/bot must be place almost at the very end of the process

Comment (memhard) [inline dockerfile]: why not just bot dir? not /app/bot

File: README.md [hunk 15]
```diff
@@ -15,67 +15,67 @@ The `Stats` command displays coin toss statistics for each user. The bot tracks
...
-   cd ..
-   docker compose up
+   docker compose up --build
 
 ___
 ## Note:
 
-- sqlite3 database is used to save usage statistics. Database folder `src/storage` 
+- sqlite3 database is used to save usage statistics. Database folder `bot/storage` 
+- logs foler `bot/storage`
```
Comment (memhard) [inline README.md]: folder?

**Review by memhard — CHANGES_REQUESTED** [review]
Need changes!

File: README.md (hunk 15 above)
Comment (Olderestin) [reply README.md]: Yes

File: bot/db.py (hunk 5 above)
Comment (Olderestin) [reply bot/db.py]: I have to describe what contains in result parameter?

File: bot/db.py (hunk 3 above)
Comment (Olderestin) [reply bot/db.py]: I have seen such description in fastapi_users library, so i thought to do it like this. As i understand, I have to describe variables in __init__

**Review by Olderestin — COMMENTED** [review]
Task done.

File: bot/config.py [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+        """
+        Configuration class for settings.
+        """
+
+        env_file = ".env"
+        case_sensitive = True
+
+settings = Settings()
+
+USER_BD: UserDatabase = UserDatabase(settings.DATABASE_PATH)
```
Comment (memhard) [inline bot/config.py]: what is BD? why we use annotation there?

File: bot/db.py [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+        Initializes the UserDatabase instance.
+
+        param db_path: The path to sqlite3 database file.
+
+        attribute db_path: The path to connect to sqlite3 database.
+        """
+        self.db_path = db_path
+        self.conn = sqlite3.connect(self.db_path / "user.db")
+
+    def __enter__(self) -> UserDatabase:
```
Comment (memhard) [inline bot/db.py]: it's doesn't return UserDatabase object + bad docstring

File: bot/db.py [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+        """
+
+        self.cursor.execute('INSERT OR IGNORE INTO user (user_id) VALUES (?)', (user_id,))
+
+    def make_flip(self, user_id: int) -> str:
+        """
+        Simulates a coin flip, updates user statistics, and returns the result.
+        
+        param user_id: The ID of the user making the flip.
+        """
```
Comment (memhard) [inline bot/db.py]: consider the possibilty to remake docstring, like example:
```bash
"""
Simulates a coin flip, updates user statistics.

Args:
      param user_id: The ID of the user making the flip.
Returns:
      some result(example)
 """
````

File: bot/db.py [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+        return result
+
+    def get_stats(self, user_id: int) -> Tuple[int, int, int]:
+        """
+        Retrieves user statistics from the database and returns it.
+        
+        param user_id: The ID of the user.
+        
+        return: tuple (flips_count, heads_count, tails_count).
+        """
```
Comment (memhard) [inline bot/db.py]: same

File: bot/handlers.py [hunk 20]
```diff
@@ -12,27 +12,48 @@
 MARKUP = ReplyKeyboardMarkup(reply_keyboard, one_time_keyboard=False, resize_keyboard=True)
 
 async def start(update: Update, context: CallbackContext) -> None:
+    """
+    Handles the /start command and initializes a new user.
+
+    param update: The incoming update.
+    param context: The context for the callback.
+    """
```
Comment (memhard) [inline bot/handlers.py]: same

File: bot/handlers.py [hunk 21]
```diff
@@ -12,27 +12,48 @@
...
     logger.info(f"Create a new user, his id: {user_id}")
     await update.message.reply_text("Привет! Теперь вы можете использовать команды flip и stats.", reply_markup=MARKUP)
 
 async def flip(update: Update, context: CallbackContext) -> None:
+    """
+    Handles the «Flip» command and simulates a coin flip for the user and send the result.
+
+    param update: The incoming update.
+    param context: The context for the callback.
+    """
```
Comment (memhard) [inline bot/handlers.py]: same

File: bot/handlers.py [hunk 22]
```diff
@@ -12,27 +12,48 @@
...
     logger.info(f"User {user_id} make a flip with result: {result}")
     await update.message.reply_text(f'Результат подбрасывания монетки: {result}')
 
 async def stats(update: Update, context: CallbackContext) -> None:
+    """
+    Handles the «Stats» command and displays the user's coin flip statistics.
+
+    param update: The incoming update.
+    param context: The context for the callback.
+    """
```
Comment (memhard) [inline bot/handlers.py]: same

File: bot/handlers.py [hunk 23]
```diff
@@ -54,6 +75,12 @@ async def stats(update: Update, context: CallbackContext) -> None:
         await update.message.reply_text('Вы еще не подбрасывали монетку. Используйте Flip.')
 
 async def about(update: Update, context: CallbackContext) -> None:
+    """
+    Handles the «About» command and provides information about the bot.
+
+    param update: The incoming update.
+    param context: The context for the callback.
+    """
```
Comment (memhard) [inline bot/handlers.py]: same

**Review by memhard — COMMENTED** [review]
LGTM 👍 but still need some changes

**Review by Olderestin — COMMENTED** [review]
Changes done.

File: bot/db.py [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+        result = 'heads' if random.randint(0, 1) == 0 else 'tails'
+        
+        if result == 'heads':
+            self.cursor.execute('UPDATE user SET heads_count = heads_count + 1 WHERE user_id = ?', (user_id,))
+        else:
+            self.cursor.execute('UPDATE user SET tails_count = tails_count + 1 WHERE user_id = ?', (user_id,))
+
+        return result
+
+    def get_stats(self, user_id: int) -> Tuple[int, int, int]:
```
Comment (memhard) [inline bot/db.py]: consider the possibility don't using typing, we can just use tuple(python 3.11)

**Review by memhard — APPROVED** [review]
LGTM 👍 can approve to the dev branch

**Pull request merged**